arrows/klv/klv_1108.cxx

```diff
@@ -0,0 +1,448 @@
+    klv_read_int< uint16_t >( checksum_it, checksum_length );
+
+  // Compare against checksum of rest of KLV packet
+  auto const calculated_checksum =
+    klv_crc_16_ccitt( data, checksum_value_begin );
+  if( calculated_checksum != checksum_value )
+  {
+    std::stringstream ss;
+    ss << std::hex << std::setfill( '0' );
+    ss  << "calculated checksum "
+        << "(0x" << std::setw( 4 ) << calculated_checksum << ") "
+        << "does not equal checksum at end of packet "
+        << "(0x" << std::setw( 4 ) << checksum_value << ")";
+    VITAL_THROW( kv::metadata_exception, ss.str() );
+  }
+
+  // Read rest of packet as normal
+  auto const result =
+    klv_local_set_format::read_typed( data, length - checksum_packet_length );
+
+  // Ensure iterator ends up after checksum packet
+  data += checksum_packet_length;
+
+  return result;
+}
+
+// ----------------------------------------------------------------------------
+void
+klv_1108_local_set_format
+::write_typed( klv_local_set const& value,
+               klv_write_iter_t& data, size_t length ) const
+{
+  // Write rest of packet as normal
+  auto const begin = data;
+  auto const internal_length = length - checksum_packet_length;
+  klv_local_set_format::write_typed( value, data, internal_length );
+
+  // Write checksum key and length, which will be included in the checksum
+  // calculation
+  klv_write_int< uint8_t >( KLV_1108_CHECKSUM, data, 1 );
+  klv_write_int< uint8_t >( 2, data, 1 );
+
+  // Write checksum value
+  auto const checksum = klv_crc_16_ccitt( begin, data );
+  klv_write_int( checksum, data, 2 );
+}
+
+// ----------------------------------------------------------------------------
+size_t
+klv_1108_local_set_format
+::length_of_typed( klv_local_set const& value, size_t length_hint ) const
+{
+  return klv_local_set_format::length_of_typed(
+    value, length_hint - checksum_packet_length ) + checksum_packet_length;
+}
+
+// ----------------------------------------------------------------------------
+std::string
+klv_1108_local_set_format
+::description() const
+{
+  return "ST 1108 local set of " + length_description();
+}
+
+// ----------------------------------------------------------------------------
+klv_uds_key
+klv_1108_key()
+{
+  // From Section 6 of https://gwg.nga.mil/misb/docs/standards/ST1108.3.pdf
+  return { 0x060E2B3402030101, 0x0E0103031C000000 };
+}
+
+// ----------------------------------------------------------------------------
+klv_tag_traits_lookup const&
+klv_1108_traits_lookup()
+{
+  // From Table 1 of https://gwg.nga.mil/misb/docs/standards/ST1108.3.pdf
+  // Descriptions are edited for clarity, brevity, consistency, etc.
+  static klv_tag_traits_lookup const lookup = {
+#define ENUM_AND_NAME( X ) X, #X
+    { {},
+      ENUM_AND_NAME( KLV_1108_UNKNOWN ),
+      std::make_shared< klv_blob_format >(),
+      "Unknown",
+      "Unknown tag.",
+      0 },
+    { // "Key" column
+      { 0x060E2B3401010101, 0x0E01050100000000 },
```

Comment:
See comment above on substituting a #define.  Applies to subsequent usage below as well.

Reply:
> The table on page 3 of the MISB standard document gives reasonable names for theses keys.

Yes, and those reasonable names are included a few lines below each key.

Each 16-byte key (composed here of two 8-byte integers) is unique, and will only appear once in the code - here, accompanied by its name, enumeration, format, and description. All future uses of this key will reference the value here (same as the local set key above - that's the reason why there is a function specifically to return that value). It seems to me that `#define`s or `constexpr` variables will simply require adding hundreds more lines of code without any substantial benefit. Their names and where their arbitrary-seeming value came from are already documented in the file. No more information is added by giving each half of each of the 200ish eventual keys its own name, to be used once. Sort of like:
```
// Add 1 for null character
auto const string_length = length + 1;
```
versus
```
constexpr size_t null_character_length = 1;
auto const string_length = length + null_character_length;
```
 The second is not obviously better than the first. Am I missing something here?

Reply:
Fair enough.  I understand that adding a few hundred lines of code with no effect on performance has a real cost to the project.  I'll leave it to @mleotta to decide if this is how project $ should be spent.